lightning-invoice/src/payment.rs

```diff
@@ -244,140 +255,178 @@ where
+						// This may happen if we send a payment and some paths fail, but
+						// only due to a temporary monitor failure or the like, implying
+						// they're really in-flight, but we haven't sent the initial
+						// HTLC-Add messages yet.
+						Ok(payment_id)
+					}
+				},
+			},
+		}.map_err(|e| PaymentError::Sending(e))
 	}
 
-	fn retry_payment(&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters)
-	-> Result<(), ()> {
-		let route;
-		{
-			let mut payment_cache = self.payment_cache.lock().unwrap();
-			let entry = loop {
-				let entry = payment_cache.entry(payment_hash);
-				match entry {
-					hash_map::Entry::Occupied(_) => break entry,
-					hash_map::Entry::Vacant(entry) => entry.insert(0),
-				};
-			};
-			if let hash_map::Entry::Occupied(mut entry) = entry {
-				let max_payment_attempts = self.retry_attempts.0 + 1;
-				let attempts = entry.get_mut();
-				*attempts += 1;
-
-				if *attempts >= max_payment_attempts {
-					log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				} else if has_expired(params) {
-					log_trace!(self.logger, "Invoice expired for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				}
+	fn retry_payment(
+		&self, entry: &mut PaymentEntry, payment_id: PaymentId, params: &RouteParameters
+	) -> Result<(), ()> {
+		let payment_hash = *entry.key();
+		let max_payment_attempts = self.retry_attempts.0 + 1;
+		let attempts = entry.get_mut();
+		*attempts += 1;
+
+		if *attempts >= max_payment_attempts {
+			log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
+		}
 
-				let payer = self.payer.node_id();
-				let first_hops = self.payer.first_hops();
-				route = self.router.find_route(&payer, &params, Some(&first_hops.iter().collect::<Vec<_>>()), &self.scorer.lock());
-				if route.is_err() {
-					log_trace!(self.logger, "Failed to find a route for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-					return Err(());
-				}
-			} else {
-				unreachable!();
-			}
+		if has_expired(params) {
+			log_trace!(self.logger, "Invoice expired for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
+		}
+
+		let payer = self.payer.node_id();
+		let first_hops = self.payer.first_hops();
+		let route = self.router.find_route(&payer, &params, Some(&first_hops.iter().collect::<Vec<_>>()), &self.scorer.lock());
+		if route.is_err() {
+			log_trace!(self.logger, "Failed to find a route for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+			return Err(());
 		}
 
-		let retry_res = self.payer.retry_payment(&route.unwrap(), payment_id);
-		match retry_res {
+		match self.payer.retry_payment(&route.unwrap(), payment_id) {
```

Comment:
After staring at it a bit more, I'm not a huge fan of calling this or `send_payment` with the retry counter lock held. I believe we currently only hold the lock during `send_payment`, and ideally we'd avoid it for both.

Reply:
That's fair. Changed to only lock as needed.